core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
-                return { requested_work, ports_status.in_samples, success ? work::Status::DONE : work::Status::ERROR };
+        } else { // block does not define any valid processing function
+            static_assert(gr::meta::always_false<gr::traits::block::stream_input_port_types_tuple<Derived>>, "neither processBulk(...) nor processOne(...) implemented");
+        }
+        forwardTags();
+        if (lifecycle::isShuttingDown(this->state()) || nextEosTag <= processedIn + 1) {
+            if (auto e = this->changeStateTo(lifecycle::State::REQUESTED_STOP); !e) {
+                emitErrorMessage("isShuttingDown -> STOPPED", e.error());
             }
-
-            // return { requested_work, ports_status.in_samples, success ? (lifecycle::isShuttingDown(std::atomic_load_explicit(&state, std::memory_order_acquire)) ? work::Status::DONE :
-            // work::Status::OK) : work::Status::ERROR };
-            return { requested_work, ports_status.in_samples, success ? work::Status::OK : work::Status::ERROR };
-        } // processOne(...) handling
-        else {
-            static_assert(gr::meta::always_false<Derived>, "neither processBulk(...) nor processOne(...) implemented");
+            updateInputAndOutputTags(processedIn);
+            applyChangedSettings();
+            ret       = work::Status::DONE;
+            processedIn = 0UZ;
         }
-        return { requested_work, 0UZ, work::Status::ERROR };
-    } // end: work_return_t work_internal() noexcept { ..}
+        // publish/consume
+        publishSamples(processedOut, outputSpans);
+        const auto inputSamplesToSkipAfter = resampledIn > 0 ? inputSamplesToSkipAfterChunk(maxSyncAvailableIn - inputSkipBefore) : 0UZ;
+        bool       success;
+        if (inputSamplesToSkipAfter > 0 && inputSamplesToSkipAfter <= resampledIn) {
+            updateInputAndOutputTags(inputSamplesToSkipAfter); // apply all tags in the skipped data range
+            success = consumeReaders(inputSamplesToSkipAfter, inputSpans);
+        } else if (inputSamplesToSkipAfter > 0) {
+            updateInputAndOutputTags(inputSamplesToSkipAfter); // apply all tags in the skipped data range
+            success                          = consumeReaders(resampledIn, inputSpans);
+            const auto inputSpansForSkipping = prepareStreams(inputPorts<PortType::STREAM>(&self()), inputSamplesToSkipAfter - resampledIn); // only way to consume is via the ConsumableSpan now
+            consumeReaders(inputSamplesToSkipAfter - resampledIn, inputSpansForSkipping);
+        } else {
+            success = consumeReaders(processedIn, inputSpans);
+        }
+        // if the block state changed to DONE, publish EOS tag on the next sample
+        if (ret == work::Status::DONE) {
+            this->setAndNotifyState(lifecycle::State::STOPPED);
+            publishTag({ { gr::tag::END_OF_STREAM, true } }, 1);
+            ret = work::Status::DONE;
```

Comment:
This line should be necessary (ret already DONE)

Reply:
there probably was a second condition in the if before... can be removed when fixing the eos behavior for process bulk that doesn't consume anything